### DISCUSSION TITLE
Adding realistic SMMU invalidation delays to gem5

### DISCUSSION THREAD

**Discussion by sammy17** [gem5/gem5#933] [Q&A] [answered]
I noticed current IOTLB invalidation in ARM SMMU is performed without any delay in the timing mode. 
I wanted to know/discuss two things:
1. What would be the best way to add the delays to the invalidation functions? 
2. Are there any realistic IOTLB invalidation delay values available from real ARM systems that gem5 simulates?

For the first point, I had some idea of where we can add the delays. 
I thought of editing the below function to pass the SMMUProcess and Yeild to the invalidation functions:
https://github.com/gem5/gem5/blob/e8bc4fc137a5a7e82b601432271a027b652ae69b/src/dev/arm/smmu_v3.cc#L388C1-L389C47
```
void
SMMUv3::processCommand(SMMUProcess::Yield &yield, const SMMUCommand &cmd, SMMUProcess *proc)
```
Then add delays around the invalidation functions as below:
```
proc->doSemaphoreDown(yield, tlbSem);
proc->doDelay(yield, Cycles(10));
tlb.invalidateVA(addr, cmd.dw0.asid, cmd.dw0.vmid);
proc->doSemaphoreUp(tlbSem);
```
Is this the correct way to add delays? Are there better ways to do this? 

Now to the second point, I wanted to know whether there are already calculated IOTLB invalidation delay values available (from experiments with real ARM SMMU devices). 
My goal is to have a more realistic system simulation with IOTLB invalidation latency values that are close to actual ARM SMMU latency. 

Appreciate any thoughts/comments about these points @giactra 
Thank you in advance!

**Comment by giactra** [comment, marked as answer]
I am afraid it is more complex than that.
The command/invalidation logic in the gem5 SMMU contains the assumption commands will be parsed and executed sequentially (with zero delay as you said) and they will be visible by other observers as soon as they are consumed. The SMMUv3 architecture does not require this serializing behaviour. Commands could be executed concurrently and completion/visibility is only guaranteed by the completion of a CMD_SYNC command.

So simply adding delays to different invalidation commands does not necessarily mean a more realistic timing model.

**Reply by sammy17** [reply]
I see, so invalidations should be scheduled like SMMUProcesses and CMD_SYNC should wait for all the processes triggered from the CMDs to be done. 
I will try to come up with something similar as time permits. 

By "other observers", did you mean ongoing translations or something else?

**Reply by giactra** [reply]
> I see, so invalidations should be scheduled like SMMUProcesses and CMD_SYNC should wait for all the processes triggered from the CMDs to be done.

Yes, correct

> By "other observers", did you mean ongoing translations or something else?

Yes what I meant is that as soon as the command is consumed (CONS is updated), the invalidation is visible by new translation requests (meaning they won't be using old data). Please have a look at the SMMUv3 documentation to have a more informed, less informal explanation on how things work ;)

**Reply by sammy17** [reply]
Thanks @giactra , I read the relevant parts of the spec and will try to implement the command consuming process in gem5. 
I have a few questions related to the ARM SMMU IOTLB invalidation delay (more related to actual ARM HW rather than gem5).
I am hoping you can help answer these questions. 
I observed that the IOTLB invalidation delay (including the CMD_SYNC process that you mentioned) can be pretty high (because of this, Linux has the option to defer IOTLB invalidation). 

**Do you know what is the cause for this high latency?** I could not figure this out based on the information given in the spec. 
Is this due to IOTLB invalidation waiting for the ongoing page table walks (that will be affected by the invalidation) to finish? 
Or are there other factors that contribute to this high latency? 
Thank you in advance for any help!

**Comment by giactra** [comment]
> Is this due to IOTLB invalidation waiting for the ongoing page table walks (that will be affected by the invalidation) to finish?

That for sure is a problem: as soon as the SMMU consumes a CMD_SYNC, it needs to wait for all pending translations to complete so that they do not see inconsistent data.

>  Or are there other factors that contribute to this high latency?

I honestly don't know. Also please bear in mind the SMMU in gem5 is an architectural model which has no claim to closely match the timings of a realistic IOMMU

**Reply by sammy17** [reply]
Thanks for the comment!
**Side note:**
Also, I noticed that the current SMMU implementation in gem5 updates the CONs before processing the commands.
According to the spec (and my understanding of it), CONs should be updated after HW consumes the command. 
Currently, this does not make any difference because commands are zero-latency processes.
In my case, I added latencies to invalidation commands so by moving the CONs update after the processCommand call, I was able to get more realistic time values.
